File: 200-number-of-islands/number-of-islands.py
def dfs(i,j,grid):
    n,m=len(grid),len(grid[0])
    grid[i][j]="0"
    dir=[(-1,0),(1,0),(0,1),(0,-1)]
    for (x,y) in dir:
        ni=i+x
        nj=j+y
        if ni>=0 and ni<n and nj>=0 and nj<m and grid[ni][nj]!="0":
            dfs(ni,nj,grid)
    return 

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n,m=len(grid),len(grid[0])
        cnt=0
        for i in range(n):
            for j in range(m):
                if grid[i][j]=="1":
                    cnt+=1
                    dfs(i,j,grid)
        
        print(grid)
        
        return cnt
        

File: 200-number-of-islands/number-of-islands.cpp
void bfs(int st_i,int st_j,vector<vector<char>>&grid,vector<vector<int>>&vis){
    queue<pair<int,int>>q;
    int n=grid.size();
    int m=grid[0].size();
    q.push({st_i,st_j});

    vector<pair<int,int>>dir{{-1,0},{1,0},{0,-1},{0,1}};
    while(!q.empty()){
        auto [x,y]=q.front();
        q.pop();
        for(auto [i,j]:dir){
            int ni=i+x;
            int nj=j+y;
            if(ni>=0 && ni<n && nj>=0 && nj<m && !vis[ni][nj] && grid[ni][nj]=='1'){
                vis[ni][nj]=1;
                q.push({ni,nj});
            }
        }
    }
    return ;
}
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        
        int n=grid.size();
        int m=grid[0].size();

        vector<vector<int>>vis(n,vector<int>(m,0));
        int cnt=0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(!vis[i][j] && grid[i][j]=='1'){
                    cnt++;
                    bfs(i,j,grid,vis);
                }
            }
        }
        return cnt;
    }
};